### PR TITLE
[TE] rootcause analysis pipeline prototype

### PR DESCRIPTION
**Pull request by apucher**
* rootcause search framework with multiple pipelines

* manual console interface

* soft URNs for entities

* linear result aggregation

* sample implementation

    * events for time range

    * events for metric

    * metrics from metric dataset

    * metric dimensions from contribution analysis

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.linkedin.thirdeye.rootcause;
+
+public class Entity {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: javadoc

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Aggregator.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.linkedin.thirdeye.rootcause;
+
+import java.util.List;
+import java.util.Map;
+
+
+public interface Aggregator {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Aggregator.java]: what is the purpose of Aggregator, can we add javadocs to all the interface and their methods

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package com.linkedin.thirdeye.rootcause.impl;
+
+import com.linkedin.thirdeye.rootcause.Entity;
+
+
+public class BaselineEntity extends Entity {
+  public static BaselineEntity fromURN(String urn, double score) {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java]: Why is baseline an entity?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricEntity.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package com.linkedin.thirdeye.rootcause.impl;
+
+import com.linkedin.thirdeye.datalayer.dto.DatasetConfigDTO;
+import com.linkedin.thirdeye.datalayer.dto.MetricConfigDTO;
+import com.linkedin.thirdeye.rootcause.Entity;
+
+
+public class MetricEntity extends Entity {
+  public static MetricEntity fromDTO(double score, MetricConfigDTO metric, DatasetConfigDTO dataset) {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricEntity.java]: Do we want to enforce having a score to the entities at all times? Even if it was created for the search context?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class FrameworkResultUtils {
+  public static Map<EntityUtils.EntityType, Collection<Entity>> topKPerType(Collection<Entity> entities, int k) {
+    Map<EntityUtils.EntityType, Collection<Entity>> map = new HashMap<>();
+    for(Entity e : entities) {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java]: In general throughout the code, can we be more verbose about variable names? When reading new code, it avoids having to go back again and again to check what a "t" or a "k" could mean.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Framework.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class Framework {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Framework.java]: Too generic name, you might want to call it RCAFramework.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/FrameworkResult.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package com.linkedin.thirdeye.rootcause;
+
+import java.util.List;
+import java.util.Map;
+
+
+public final class FrameworkResult {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/FrameworkResult.java]: What does frameworkResult mean? Is the result of a single execution?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java (hunk 3 above)
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java]: if it's representing timerange, have an abstract class TimeRangeEntity. CurrentEntity, BaselineTimeRangeEntity can extend the TimeRangeEntity.

Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java]: Also, provide sample urn for each entity.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  public enum EntityType {
+    EVENT("thirdeye:event"),
+    METRIC("thirdeye:metric"),
+    TIMERANGE("thirdeye:timerange"),
+    BASELINE("thirdeye:baseline"),
+    UNKNOWN("");
+
+    private final String prefix;
+
+    EntityType(String prefix) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java]: why is this package level access

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java (hunk 5 above)
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java]: +1 on variable names. Please evaluate the need for a util class to achieve this functionality.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java (hunk 1 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: javadoc added

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Aggregator.java (hunk 2 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Aggregator.java]: javadoc added

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+ * Utility class to simplify type-checking and extraction of entities
+ */
+public class EntityUtils {
+  // timerange:  thirdeye:timerange:{start}:{end}
+  // baseline:   thirdeye:baseline:{start}:{end}
+  // event:      thirdeye:event:{type}:{name}:{start}
+  // metric:     thirdeye:metric:{dataset}:{name}
+  // metric dim: thirdeye:metric:{dataset}:{name}:{dimension}
+
+  public enum EntityType {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java]: move this to its own class

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java (hunk 3 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/BaselineEntity.java]: samples added.
@npawar we could have used a baseline time-range (pseudo-entity) in the search context as well. we went with an entity wrapper to stay consistent with out knowledge graph schema

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricDatsetPipeline.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Sample implementation of a pipeline for identifying relevant metrics based on dataset
+ * association. The pipeline first fetches metric entities from the context and then
+ * searches Thirdeye's internal database for metrics contained in the same datasets as
+ * any metric entities in the search context. All found metrics are scored equally.
+ */
+public class MetricDatsetPipeline implements Pipeline {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricDatsetPipeline.java]: typo datset

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricEntity.java (hunk 4 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricEntity.java]: multiple options again. Ultimately, we want weights provided by the user in the search context, and weights generated by pipelines and the aggregation stage in the execution context.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java (hunk 5 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/FrameworkResultUtils.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Framework.java (hunk 6 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Framework.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java (hunk 8 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java]: Enum constructor is always private

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java (hunk 9 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/EntityUtils.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricDatsetPipeline.java (hunk 10 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricDatsetPipeline.java]: fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/FrameworkResult.java (hunk 7 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/FrameworkResult.java]: javadoc added

**Pull request merged**